Marlin/Marlin_main.cpp

```diff
@@ -3968,6 +3982,22 @@ inline void gcode_M109() {
     if (code_seen('B')) autotemp_max = code_value();
   #endif
 
+  // Detect if a print job has finished.
+  // When the target temperature for all extruders is zero then we must have
+  // finished printing.
+  if( print_job_start_ms != 0 ) {
+    bool all_extruders_cooling = true;
+    for (int i = 0; i < EXTRUDERS; i++) if( degTargetHotend(i) > 0 ) {
+      all_extruders_cooling = false;
+      break;
+    }
+
+    if( all_extruders_cooling ) {
+      print_job_stop_ms = millis();
+      LCD_MESSAGEPGM(WELCOME_MSG);
+    }
+  }
```

Comment:
Then I'll take the opportunity to rework the overall code a bit a implement a start_print_job() function also.